Improve jetpack-mu-wpcom Translation Updates with Automatic Download & Version Checking

d-alleyne opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes Automattic/i18n-issues#875

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
 Adds a mechanism to fetch translation updates from the wp.com Translation API. It is modelled after the WooCommerce Translation Update mechanism PCYsg-sh4-p2. 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

### **Setup**  
1. Set up an Atomic Site using PCYsg-Osp-p2 (**Develop wpcomsh using a WoA dev blog site**).  

### **Verification Steps**  
2. Once the changes are deployed, SSH into the Atomic Site and verify:  
   - That the `htdocs/wp-content/languages/mu-plugins` directory exists and contains the latest translation files for `jetpack-mu-wpcom`.  
   - Check a translation file, such as:  
     ```sh  
     head htdocs/wp-content/languages/mu-plugins/jetpack-mu-wpcom-es_ES.po  
     ```  
     Take note of the `PO-Revision-Date` header value (e.g., **2025-03-05 14:54:03+0000**).  
   - Ensure the `wpcomsh_translation_update` hook exists:  
     ```sh  
     wp cron event list --hook=wpcomsh_translation_update  
     ```  

### **Override with Older Translations**  
3. Replace the current translations with older versions by modifying:  
   - **From:**  
     ```php  
     $plugins_request_data[ $plugin_slug ] = array( 'version' => 'latest' );  
     ```  
   - **To:**  
     ```php  
     $plugins_request_data[ $plugin_slug ] = array( 'version' => '2.7.0' );  
     ```  
   - **From:**  
     ```php  
     if ( $new_translation_revision_time <= $installed_translation_revision_time ) {  
         continue;  
     }  
     ```  
   - **To:** *(Comment out this condition to allow older translations to overwrite newer ones)*  
     ```php  
     // if ( $new_translation_revision_time <= $installed_translation_revision_time ) {  
     //    continue;  
     // }  
     ```  

### **Execute and Verify Update**  
4. Run the cron job to update translations:  
   ```sh  
   wp cron event run wpcomsh_translation_update  
   ```  
5. Verify that the translation file reflects an older revision date than the one recorded earlier:  
   ```sh  
   head htdocs/wp-content/languages/mu-plugins/jetpack-mu-wpcom-es_ES.po  
   ```
6. Change your user locale to a MAG-16 locale that was never used on your Atomic Site in a different tab, e.g. (Hebrew )
7. Load `/wp-admin/options-general.php` on your Atomic Site, and verify that [Site Logo](https://github.com/Automattic/jetpack/blob/be7ca778b42008b23b48987ce725d7adceb0a9f2/projects/packages/jetpack-mu-wpcom/src/features/wpcom-options-general/options-general.php#L33) is untranslated
![CleanShot 2025-03-07 at 10 43 14@2x](https://github.com/user-attachments/assets/783e2fb2-13bd-4f6a-9044-f403ba68d3c9)
8.  Execute the cronjob manually (step 4), and verify that the new locale now exists in the `htdocs/wp-content/languages/mu-plugins/` folder
9. Refresh `/wp-admin/options-general.php`, and verify that this text is now translated.
![CleanShot 2025-03-07 at 10 46 12@2x](https://github.com/user-attachments/assets/49597c47-ffee-4418-97b6-a88de15128e9)

d-alleyne committed:
Update translations on loading the plugin

d-alleyne committed:
changelog

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations

d-alleyne commented:
@anomiex, I have a few questions regarding the implementation of this feature on **Atomic sites**:  

1. How should translation updates be scheduled?  
2. What is the best destination directory for updated translations?  

I am following the approach outlined in **PCYsg-sh4-p2**, but there are key differences due to `wpcomsh` being an **`mu-plugin`**. Since `mu-plugins` do not hook into WordPress' native update system, we can’t rely on core’s built-in update mechanisms. ([Reference](https://make.wordpress.org/core/2020/07/30/recommended-usage-of-the-updates-api-to-support-the-auto-updates-ui-for-plugins-and-themes-in-wordpress-5-5/))  

### Proposed Approach:  
- Download translation files from the **WordPress.com GlotPress API** with each rolling release.  
- Store updates in `wp-content/languages/mu-plugins/`.  
- Use the existing `Scheduled_Updates` class to handle periodic refreshes.  

### Questions About Storage Location: 
I noticed that **`wp-content/languages/mu-plugins/` is not present on Atomic sites by default**. Would this be the correct location for storing `mu-plugin` translations, or would it be better to use **`wp-content/languages/plugins/`**, which WordPress already recognizes for plugin translations?  

Additionally, I saw that [wpcomsh/i18n.php](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/wpcomsh/i18n.php#L68) includes the following line:  
```php
load_muplugin_textdomain( 'wpcomsh', 'wpcomsh/languages' );
```  
This suggests that `wpcomsh` expects translations to be loaded from `wpcomsh/languages/`.  

- Would it be feasible to store and update translations in `wpcomsh/languages/` on Atomic, or is that directory protected?  
- If modifying `wpcomsh/languages/` isn’t ideal, should we stick with `mu-plugins/`, or is `plugins/` the better location?  

Any guidance would be appreciated. Thanks!

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -521,4 +522,118 @@ public static function load_social_links() {
+			),
+		);
+
+		$response = wp_remote_post(
```
Will this wind up performing this fetch on every page view on the site? Is that going to be a performance concern, for the site itself and for translate.wordpress.com?

d-alleyne replied:
The goal of this was to run the fetch once the plugin was loaded/activated, not on every page request! I can add caching to prevent this, but I think using transients with a scheduled cron job might be a better alternative.

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -521,4 +522,118 @@ public static function load_social_links() {
+	private static function install_translation_package( $package_url ) {
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+
+		$wpcomsh_dir = is_dir( WP_CONTENT_DIR . '/mu-plugins/wpcomsh-dev/' ) ? 'wpcomsh-dev' : 'wpcomsh';
```
I note it's fairly likely that files placed in `mu-plugins/wpcomsh-dev/` will get deleted. As written, the code above probably won't restore them since it'll hit it "If translation is already the latest version, skip" check.

If it'll work, it might make more sense to write into `wp-content/languages/plugins/`, like WordPress does for translations it manages, or possibly `wp-content/languages/mu-plugins/` if you want to keep a plugin/mu-plugin separation.

anomiex reviewed (commented):
Approach itself seems ok, although you might put the code into `projects/plugins/wpcomsh/i18n.php` instead if you want.

Some concerns flagged inline.

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations

d-alleyne committed:
Integrate translation updates with Scheduled_Updates

- Integrates `maybe_update_translations()` with Jetpack's `Scheduled_Updates` system.
- Change the storage location for downloaded translations to the `WP_LANG_DIR/mu-plugins`

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -521,4 +522,117 @@ public static function load_social_links() {
+		}
+
+		// Extract language files into the destination directory.
+		$unzip_result = unzip_file( $temp_file, $destination );
```
Do we need an equivalent of core's `Language_Pack_Upgrader->clear_destination()` before we unzip the file?

Or I wonder if we could reuse `Language_Pack_Upgrader` to do most of the work for us by passing it the right information. At a quick glance, maybe, although we may have to use the 'upgrader_package_options' hook to override the destination to `WP_LANG_DIR . '/mu-plugins/'` if we want that instead of `/plugins`.

d-alleyne replied:
That’s a great suggestion! Initially, I was worried that [clear_destination](https://developer.wordpress.org/reference/classes/language_pack_upgrader/clear_destination/) might delete all translation files inside `wp-content/languages/mu-plugins/`, potentially affecting other MU plugins like `wpcomsh` (when we migrate those files to this directory). However, after reviewing the actual implementation, I confirmed that it only deletes translation files specifically for the language and plugin being updated.

d-alleyne replied:
Adding `Language_Pack_Upgrader` adds unexpected dependencies, and a lot of warning messages regarding [undefined](https://github.com/WordPress/wordpress-develop/blob/8a52d746e9bb85604f6a309a87d22296ce1c4280/src/wp-admin/includes/class-language-pack-upgrader.php#L413) [properties](https://github.com/WordPress/wordpress-develop/blob/6.7/src/wp-admin/includes/class-language-pack-upgrader.php#L416). I'll implement it manually to reduce reliance on these dependencies.

anomiex reviewed (commented):
Looks reasonable. Not sure how to test.

d-alleyne committed:
Schedule daily translation updates

Jetpack's `Scheduled_Updates` does not apply to mu-plugins,
so our WPCOMSH/jetpack-mu-wpcom translations will not update as expected.

This ensures translations are updated daily **without slowing down site performance**.

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -91,6 +91,29 @@ public static function init() {
+	 * @return void
+	 */
+	public static function schedule_translation_updates() {
+		// Only schedule the cron job if it hasn't been scheduled before.
```
Is the intention to allow the site admin to unschedule the event if they don't want translation updates to ever happen?

d-alleyne replied:
No. The goal is to set up a cronjob once that will check for translation updates daily (or maybe a few times per day).

anomiex replied:
In that case, doesn't the use of `if ( ! wp_next_scheduled( 'wpcomsh_daily_translation_update' ) )` already handle it? Why the option as well?

d-alleyne committed:
Use Language_Pack_Upgrader to handle file extraction and cleanup

d-alleyne committed:
Refactor `maybe_update_translations` to support multiple plugins

d-alleyne committed:
Implement code review suggestion

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations

d-alleyne committed:
Bugfix

d-alleyne committed:
Include dependencies of Language_Pack_Upgrader

d-alleyne committed:
Update the schedule to twice daily

This matches the frequency of language updates for themes and plugins.

d-alleyne committed:
Fix PHP Stan security warnings

d-alleyne committed:
Minimize warnings at CLI

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations

d-alleyne committed:
Replace Language_Pack_Upgrader with a local equivalent

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -90,6 +92,167 @@ public static function init() {
+		if ( array_key_exists( 'success', $data ) && false === $data['success'] ) {
+			return;
+		}
+		// Fix PHP Stan error
```
We don't use phpstan in the monorepo (we use phan instead). And probably no need to point out that the check is avoiding an error.

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -90,6 +92,167 @@ public static function init() {
+			return;
+		}
+		// Fix PHP Stan error
+		if ( ! array_key_exists( 'data', $data ) && ! is_array( $data['data'] ) ) {
```
Should this be `||` rather than `&&`?

anomiex commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -65,6 +65,8 @@ public static function init() {
 		// These features run only on atomic sites.
 		if ( defined( 'IS_ATOMIC' ) && IS_ATOMIC ) {
 			add_action( 'plugins_loaded', array( __CLASS__, 'load_custom_css' ) );
+			add_action( 'init', array( __CLASS__, 'maybe_update_translations' ) );
```
I noticed the download is running on every page view. I think you forgot to delete this line when you added the scheduling.
```suggestion
```

anomiex reviewed (commented):
Seems to work, although I spotted two issues.

d-alleyne committed:
Validate data before using it

d-alleyne committed:
Update projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

d-alleyne committed:
Use load_theme_textdomain to load translations from WP_LANG_DIR/mu-plugins

During testing, I attempted to use load_muplugin_textdomain to load translations for
jetpack-mu-wpcom with the following approach:

    load_muplugin_textdomain( 'jetpack-mu-wpcom', '../languages/mu-plugins' );

However, this method relied on a relative path resolving to WP_LANG_DIR/mu-plugins.

To improve reliability, we switched to load_theme_textdomain with an absolute path:

    load_theme_textdomain( 'jetpack-mu-wpcom', WP_LANG_DIR . '/mu-plugins' );

The path corresponds to the download+extraction location of the language packs in the
Jetpack_Mu_Wpcom::maybe_update_translations() method.

d-alleyne committed:
changelog

anomiex reviewed (dismissed):
LGTM, although review from someone more familiar with i18n stuff might be useful. 🙂

As for the failing code coverage requirements, there are a few options:

* Add PHPUnit tests. 🙂
* Add https://github.com/Automattic/jetpack/labels/Coverage%20tests%20to%20be%20added%20later to get this shipped, and look at testing in a followup.
* Since it's not required, ignore it. :sweat_smile:  Or add https://github.com/Automattic/jetpack/labels/I%20don%27t%20care%20about%20code%20coverage%20for%20this%20PR to basically the same effect.

Which you choose (or come up with another option) is up to you and your team. 🙂

d-alleyne committed:
Add support for Javascript translation files

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations

d-alleyne committed:
Remove unnecessary check

These checks already occur in WordPress' `wp-includes/l10n.php`

d-alleyne committed:
Merge remote-tracking branch 'origin/trunk' into update/local-translations
